File: javascript_solutions/53-Maximum-Subarray.js
/*  
Question Link: https://leetcode.com/problems/maximum-subarray/
*/

/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    // DP
    let local_max = 0;
    let global_max = Number.NEGATIVE_INFINITY;

    for (let i = 0; i < nums.length; i++) {
        local_max = Math.max(nums[i], nums[i] + local_max)

        if (local_max > global_max) {
            global_max = local_max
        }
    }

    return global_max
};

File: python_solutions/53-maximum-subarray.py
# https://leetcode.com/problems/maximum-subarray/

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # A great video explained everything.
        # https://www.youtube.com/watch?v=2MmGzdiKR9Y
        
        for i in range(1, len(nums)):
            if nums[i-1] > 0:
                nums[i] += nums[i-1]
        return max(nums)